ER図を書く。 - READMEにデータベース設計を記述する

tubuyakisyun opened the pull request:
# what
**二回目**のコメントを修正し、ER図を書き直し、READMEにデータベース設計を記述し直しました
# why
furimaアプリの**デーでベース**を実装するため

tubuyakisyun committed:
ER図を書く。 - READMEにデータベース設計を記述する

tubuyakisyun commented:
# コメントを修正し、gyazoでdb設計図を送りました

https://gyazo.com/26d59274208dc51f1a6e319afb60d06b

###  [**２回目**修正したER図](https://gyazo.com/48c29b4f82215fd49f98779a479cfde1)

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| Column             | Type   | Options     |
+| ------------------ | ------ | ----------- |
+| nickname           | string | null: false |
+| email              | string | null: false |
```
emailカラムに、unique:trueオプションを追加しましょう。
このオプションにより、登録メールアドレスに一意性を持たせることができます。

tubuyakisyun replied:
# 修正しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| nickname           | string | null: false |
+| email              | string | null: false |
+| encrypted_password | string | null: false |
+| birth_date         | string | null: false |
```
誕生日はdate型を使用しましょう。date型は「年・月・日」を1つにまとめて保存する型となります。

tubuyakisyun replied:
# 修正しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| ------------------ | -----------| ------------------------------ |
+| name               | string     | null: false                    |
+| info               | text       | null: false                    |
+| price              | string     | null: false                    |
```
値段は数値ですので、integer型にしましょう。

tubuyakisyun replied:
# 修正しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| name               | string     | null: false                    |
+| info               | text       | null: false                    |
+| price              | string     | null: false                    |
+| sales_status       | string     | null: false                    |
```
こちらはプルダウン（ActiveHash）にて実装する箇所かと存じます。
ActiveHashの仕様に合わせ、TypeはInteger型とし、ActiveHashのidを保存するようにしましょう。 またカラム名は”単数型_id”としましょう。

詳しくは、下記補足カリキュラムをご参考ください。
 ・最終課題補足カリキュラム〜カテゴリーの選択を工夫しよう〜

他、ActiveHashを使用する箇所は、同様に修正しましょう。

tubuyakisyun replied:
### 修正しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+
+## item_details テーブル
+
+| Column              | Type   | Options     |
+| ------------------  | ------ | ----------- |
+| category            | string | null: false |
+| shipping_fee_status | string | null: false |
+| scheduled_delivery  | string | null: false |
```
こちらは出品画面で保存するデータかと存じます。
itemsテーブルでまとめて保存した方が実装しやすくなりますので、こちらの3つのカラムはitemsテーブルに含めるようにしましょう。

tubuyakisyun replied:
#### 修正しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
-* Database creation
+| Column             | Type   | Options     |
+| ------------------ | ------ | ----------- |
+| token              | string | null: false |
```
購入用のトークンは1度きりしか使わないため、保存しておく必要はありません。
削除しておきましょう。

tubuyakisyun replied:
#### 削除しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| ------------------ | ------ | ----------- |
+| token              | string | null: false |
+| postal_code        | string | null: false |
+| prefecture         | string | null: false |
```
こちらはプルダウン（ActiveHash）にて実装する箇所かと存じます。
ActiveHashの仕様に合わせ、TypeはInteger型とし、ActiveHashのidを保存するようにしましょう。 またカラム名は”単数型_id”としましょう。

また、カラム名は、商品テーブルの都道府県と同じカラム名にしましょう。
（こちらのテーブルのカラム、商品テーブルのカラムのどちらを変更していただいてもOKです）
ActiveHashのデータは使い回すことが可能です。 商品テーブルと配送先テーブルの「都道府県」表すカラムは、同じカラム名とすると、実装しやすくなります。 

tubuyakisyun replied:
#### カラム名は”単数型_id”としました、商品テーブルと配送先テーブルの「都道府県」表すカラムは、同じカラム名に修正しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+| postal_code        | string | null: false |
+| prefecture         | string | null: false |
+| city               | string | null: false |
+| address            | string | null: false |
```
「建物名」を保存するカラムがないかと存じます。追加しましょう。

tubuyakisyun replied:
#### building_name というカラムを追加しました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
-* ...
+### Association
+
+- belongs_to :users
```
belongs_toの関係ですので、単数形としましょう。機能実装時にエラーが発生してしまいます。

tubuyakisyun replied:
#### belongs_toの関係の時、単数形にしました

AsukaToyoshima commented on README.md:
```diff
@@ -1,24 +1,64 @@
+- has-one :purchases
 
-* Configuration
+## purchases テーブル
```
配送先の情報とは別に、
購入履歴として、「誰が」「何を」購入したかを管理するテーブルを作成しましょう。
分離することで、後々、実装がしやすくなります。

tubuyakisyun replied:
#### 購入履歴として、「誰が」「何を」購入したかを管理するテーブルを作成しました

AsukaToyoshima reviewed (commented):
レビューのご依頼ありがとうございます。
10点コメントさせていただきました。ご確認の上、修正をお願いいたします。

コメントの修正により、ER図も変更になるかと存じます。
併せて修正していただき、本プルリクエストに添付をお願いいたします。

疑問点・不明点等ございましたら、コネクトメンターへお気軽にご質問ください。

tubuyakisyun committed:
ER図を書く。 - READMEにデータベース設計を記述する

tubuyakisyun committed:
ER図を書く。 - READMEにデータベース設計を記述するレビュー：２回目

tsuyoshimizu commented on README.md:
```diff
@@ -1,24 +1,87 @@
+| ---------------------- | ---------- | ------------------------------   |
+| id                     | integer    | null:   false                    |
+| nickname               | string     | null:   false                    |
+| email                  | string     | unique:  true                    |
```
```
| email                  | string     | unique:  true                    |
```
こちら```unique: true```を付けていただいたことにより```null: false```が消されているようです。
どちらも必要になりますので
```null: false, unique: true```というようにどちらも記述しましょう。

tubuyakisyun replied:
#### null: false, unique: trueというように修正しました

tsuyoshimizu commented on README.md:
```diff
@@ -1,24 +1,87 @@
+
+| Column             | Type       | Options                        |
+| ------------------ | -----------| ------------------------------ |
+| buyer              | string     | null: false                    |
```
```
| buyer | string | null: false |
```
こちら不要なため削除しましょう。

理由：購入記録テーブルに誰が購入したかはuserで確認がとれるためです。

tubuyakisyun replied:
#### buyerカラムを削除しました

tubuyakisyun replied:
# what
**二回目**のコメントを修正し、ER図を書き直し、READMEにデータベース設計を記述し直しました
# why
furimaアプリの**デーでベース**を実装するため

###  [**２回目**修正したER図](https://gyazo.com/48c29b4f82215fd49f98779a479cfde1)

tsuyoshimizu reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
コメントいたしましたので、ご確認いただけますと幸いです。
修正後、コミットプッシュを確認いただき、再度**レビュー依頼フォーム**のご提出をお願いいたします。
お手数をお掛け致しますが、よろしくお願いいたします。

tubuyakisyun committed:
email カラムにnull: false, unique: trueというようにどちらも記述しました

tubuyakisyun committed:
buyer カラムを削除しました

理由：購入記録テーブルに誰が購入したかはuserで確認がとれるためです。

tubuyakisyun committed:
buyer　カラムを削除しました

tubuyakisyun committed:
purchases のbuyer カラムを削除しました

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,86 @@
 
-* Configuration
+## user_delivery_addresses テーブル
 
-* Database creation
+| Column               | Type       | Options                        |
+| -------------------- | ---------- | ------------------------------ |
+| user                 | references | null: false, foreign_key: true |
+| delivery_address     | references | null: false, foreign_key: true |
```
中間テーブルは不要になるため削除をお願いいたします。

「理由」
中間テーブルを使用して実装も可能ですが、カリキュラムの内容と実装のやり方が変わってしまうため、今回は使用しない形になります。
具体的には「フォームオブジェクト」というものを使用し、一度に「購入管理テーブル」と「配送先テーブル」にデータを保存する実装を行うため、削除いただくようお願いいたします。

またテーブルの関係性ですが、delivery_addressesテーブルはpurchasesテーブルとのみ紐づく形になるため、アソシエーションの修正もお願いいたします。

tubuyakisyun replied:
- 中間テーブルは不要のため、削除しました
- delivery_addressesテーブルはpurchasesテーブルとのみ紐づくアソシエーションを修正し、belongs_to :purchaseとしました

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,86 @@
-* Deployment instructions
+| Column                 | Type       | Options                        |
+| ---------------------- | ---------- | ------------------------------ |
+| postal_code_id         | integer    | null: false                    |
```
`| postal_code_id         | integer    | null: false                    |`
郵便番号ですが「ハイフン」が入ることを考慮し、string型に変更しましょう！
またデータはidではないため、カラム名の「_id」の削除をお願いいたします。

tubuyakisyun replied:
- postal_code カラム名を修正しました
- postal_code カラム型を修正しました

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| postal_code_id         | integer    | null: false                    |
+| prefecture_id          | integer    | null: false                    |
+| city                   | string     | null: false                    |
+| building_name          | string     | null: false                    |
```
`| building_name          | string     | null: false                    |`
マンションなどの建物名は、入力が任意の箇所になりますので
null :falseオプションは削除しましょう！

tubuyakisyun replied:
#### null :falseオプションは削除しました

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,86 @@
+| building_name          | string     | null: false                    |
+| address                | string     | null: false                    |
+| phone_number           | integer    | null: false                    |
+| item                   | references | null: false, foreign_key: true |
```
`| item                   | references | null: false, foreign_key: true |`
配送先テーブルは購入管理テーブルと紐づくため、itemの外部キーではなく、purchaseの外部キーの記述をお願いいたします。

tubuyakisyun replied:
#### 配送先テーブルは購入管理テーブルと紐づくため、購入管理の外部キーのpurchaseを記述し直しました。

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,86 @@
+
+### Association
+
+- belongs_to :item
+- has_many :user_delivery_addresses
+- has_many :users, through: :user_delivery_addresses
```
アソシエーションも同様に、購入管理テーブルのアソシエーションのみになるよう修正をお願いいたします。

tubuyakisyun replied:
#### delivery_addressesテーブルはpurchasesテーブルとのみ紐づくアソシエーションを修正し、belongs_to :purchaseとしました

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,86 @@
+### Association
+
+- belongs_to :user
+- has_many :delivery_addresses
```
`has_many :delivery_addresses`
商品は配送先情報を持たないため、アソシエーションの削除をお願いいたします。

tubuyakisyun replied:
#### has_many :delivery_addresses のアソシエーションを削除しました

Saka-toku commented on README.md:
```diff
@@ -1,24 +1,86 @@
+### Association
+
+- belongs_to :item
+- belongs_to :user
```
購入管理テーブルのアソシエーションも追記をお願いいたします。

tubuyakisyun replied:
#### 購入管理テーブルのアソシエーションも追記をしました

tubuyakisyun replied:
### コメントして頂いて、ありがとうございます！以上は修正の内容でした。

# what
**３回目**のコメントを修正し、ER図を書き直し、READMEにデータベース設計を記述し直しました
# why
furimaアプリの**デーでベース**を実装するため

###  [**３回目**修正したER図](https://gyazo.com/3b828054b132d1d83107d7aa4f89c4df)

Saka-toku reviewed (commented):
レビュー依頼ありがとうございます！
修正点を確認させていただきました。

カリキュラムの実装と合わせるため、いくつかコメントしております。
ご確認お願いいたします！

tubuyakisyun committed:
delivery_addresses テーブルにbuilding_name カラムを追加する

tubuyakisyun committed:
delivery_addressesテーブルはpurchasesテーブルとのみ紐づく形になるアソシエーションを修正しました

中間テーブルは不要になるため削除
「理由」
中間テーブルを使用して実装も可能ですが、カリキュラムの内容と実装のやり方が変わってしまうため、今回は使用しない形になります。
具体的には「フォームオブジェクト」というものを使用し、一度に「購入管理テーブル」と「配送先テーブル」にデータを保存する実装を行うため

tubuyakisyun committed:
郵便番号をstring型に変更 またデータはidではないため、カラム名の「_id」の削除する

郵便番号ですが「ハイフン」が入ることを考慮し、string型に変更しましょう！
またデータはidではないため、カラム名の「_id」の削除をお願いいたします。

tubuyakisyun committed:
マンションなどの建物名は、入力が任意の箇所になりますので null :falseオプションは削除

tubuyakisyun committed:
配送先テーブルは購入管理テーブルと紐づくため、購入管理の外部キーのpurchaseを記述し直しました

tubuyakisyun committed:
furima_35933のER図を更新した５回目

yuri240 commented on README.md:
```diff
@@ -1,24 +1,69 @@
-Things you may want to cover:
+| Column                 | Type       | Options                          |
+| ---------------------- | ---------- | ------------------------------   |
+| id                     | integer    | null:   false                    |
```
```
| id                     | integer    | null:   false                    |
```
idカラムは不要と思われますので、
削除しておきましょう！

tubuyakisyun replied:
#### ユーザー　id を削除しました

yuri240 commented on README.md:
```diff
@@ -1,24 +1,69 @@
+
+- belongs_to :item
+- belongs_to :user
+- has_many :delivery_addresses
```
```
- has_many :delivery_addresses
```
上記のアソシエーションですが、
1対1を示す、`has_one`に変更しましょう！

tubuyakisyun replied:
#### 1対1を示す、has_oneに変更しました

tubuyakisyun replied:
### コメントして頂いて、ありがとうございます！以上は修正の内容でした。

# what
**４回目**のコメントを修正し、ER図を書き直し、READMEにデータベース設計を記述し直しました
# why
furimaアプリの**デーでベース**を実装するため

###  [**４回目**修正したER図](https://gyazo.com/77e032d0dc0980b9f292eb7fdda524dc)

yuri240 reviewed (commented):
レビュー依頼ありがとうございます！
キャプチャ画像と修正点確認させていただきました。

追加で２点コメントさせていただきました。

あと少しかと思いますので
もう一息、頑張りましょう！

tubuyakisyun committed:
user id カラムを削除しました

tubuyakisyun committed:
has_many :delivery_addressesをhas_one :delivery_addressへ変更

Hide-Morishita commented on README.md:
```diff
@@ -1,24 +1,68 @@
+| city                   | string     | null: false                    |
+| building_name          | string     |                                |
+| address                | string     | null: false                    |
+| phone_number           | integer    | null: false                    |
```
```ruby
| phone_number | integer | null: false |
```
電話番号ですが、integer型に指定すると仮に「09012345678」を保存しようとした場合、先頭の0が保存されないので、string型に修正しましょう。

tubuyakisyun replied:
# what
- 電話番号の型：string型に変更しました
- READMEの内容を最新に反映させました

# why
furima アプリdb設計を実装するため

- ER図は以下になります
 [最新のER図](https://gyazo.com/ea59921c9356dd70a5ca9a3154e113e1)

tubuyakisyun replied:
# what
- ER図の外部キー型をreferencesに変更

# why
furima アプリdb設計を実装するため

- ER図は以下になります
 [最新のER図](https://gyazo.com/470e646514dcefdac8ef15fb12c0b609)

Hide-Morishita reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
追加でコメントいたしましたので、ご確認お願いいたします。
コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りいただくようお願いいたします。

次回レビュー依頼の際に、READMEの内容を反映させた、最新のER図の提出をお願いいたします。

tubuyakisyun committed:
電話番号の型：string型に変更し、READMEの内容を反映させました

電話番号ですが、integer型に指定すると仮に「09012345678」を保存しようとした場合、先頭の0が保存されないので、string型に修正します

Hide-Morishita commented:
レビュー依頼ありがとうございます。
修正点確認させていただきました。
ER図に関してですが、各テーブルの外部キーがREADEMEはreferences型で記述されていますが、ER図ではinteger型で記述されています。
お手数をおかけいたしますが、**READMEの内容に合わせてER図の修正**をお願いいたします。
修正後再度レビュー依頼をお送りいただくようお願いいたします。

tubuyakisyun committed:
ERずにある外部キーの型：references型に変更しました

tubuyakisyun committed:
Revert "電話番号の型：string型に変更し、READMEの内容を反映させました"

This reverts commit 9287184c15eb326e01a80e3a60f13f3d6680d6f2.

tubuyakisyun committed:
phone_number の型をもう一回変更

tubuyakisyun committed:
referencesのスペルミスを修正

yuri240 commented:
修正とレビュー依頼ありがとうございます。

ER図も拝見いたしました。

しっかりと設計できていることが確認できましたので、
**LGTM**とさせていただきます！

こちらのブランチは忘れずにMergeを行っていただき、
ローカルのMasterにPullしておきましょう。

データベース設計の実装、大変お疲れ様でした。
引き続き、応援しております。
